packages/flutter/lib/src/widgets/hero_controller.dart

```diff
@@ -10,68 +10,46 @@ import 'framework.dart';
 import 'heroes.dart';
 import 'navigator.dart';
 import 'overlay.dart';
-import 'page.dart';
-
-class HeroPageRoute extends PageRoute {
-  HeroPageRoute({
-    WidgetBuilder builder,
-    NamedRouteSettings settings: const NamedRouteSettings(),
-    this.heroController
-  }) : super(builder: builder, settings: settings);
-
-  final HeroController heroController;
-  NavigatorState _navigator;
-
-  void didPush(OverlayState overlay, OverlayEntry insertionPoint) {
-    super.didPush(overlay, insertionPoint);
-    // TODO(abarth): Pass the NavigatorState explicitly.
-    if (overlay != null) {
-      _navigator = Navigator.of(overlay.context);
-      heroController?.didPush(_navigator, this);
-    }
-  }
-
-  void didPop(dynamic result) {
-    super.didPop(result);
-    if (_navigator != null) {
-      heroController?.didPop(_navigator, this);
-      _navigator = null;
-    }
-  }
-}
+import 'routes.dart';
 
-class HeroController {
+class HeroController extends NavigatorObserver {
   HeroController() {
     _party = new HeroParty(onQuestFinished: _handleQuestFinished);
   }
 
   HeroParty _party;
   PerformanceView _performance;
-  HeroPageRoute _from;
-  HeroPageRoute _to;
+  ModalRoute _from;
+  ModalRoute _to;
 
   final List<OverlayEntry> _overlayEntries = new List<OverlayEntry>();
 
-  void didPush(NavigatorState navigator, HeroPageRoute route) {
+  void didPushModal(Route route) {
```

Comment:
It's confusing that didPushModal can be called with something that isn't a ModalRoute.  Maybe add a TODO?